gurmukhiutils/unicode.py

```diff
@@ -0,0 +1,448 @@
+        The same string with gurmukhi diacritics arranged in a sorted manner.
+
+    Example:
+        >>> sort_diacritics("\u0a41\u0a4b") == "\u0a4b\u0a41"  # ੁੋ vs  ੋੁ
+        True
+    """
+
+    import re
+
+    """
+    Nukta is essential to form a new base letter and must be ordered first.
+
+    Udaat, Yakash, and subjoined letters should follow.
+
+    Subjoined letters are constructed (they are not single char), so they cannot be used in the same regex group pattern. See further below for subjoined letters.
+    """
+    nukta_udaat_yakash_order_list = [
+        "਼",
+        "ੑ",
+        "ੵ",
+    ]
+
+    """
+    More generally, when a consonant or independent vowel is modified by multiple vowel signs, the sequence of the vowel signs in the underlying representation of the text should be: left, top, bottom, right.
+
+    p. 491 of The Unicode® Standard Version 14.0 – Core Specification
+
+    https://www.unicode.org/versions/Unicode14.0.0/ch12.pdf
+    """
+    vowel_order_list = [
+        "ਿ",
+        "ੇ",
+        "ੈ",
+        "ੋ",
+        "ੌ",
+        "ੁ",
+        "ੂ",
+        "ਾ",
+        "ੀ",
+    ]
+
+    """
+    The remaining diacritics are to be sorted at the end according to the following order
+    """
+    remaining_diacritic_order_list = [
+        "ਁ",
+        "ੱ",
+        "ਂ",
+        "ੰ",
+        "ਃ",
+    ]
+
+    """
+    If subjoined were single code points, we could have done a simple regex match:
+    ([  list_of_diacritics  ]+)
+
+    Since otherwise, surrounds each lookup of a subjoined letter with lookups of the rest of the diacritics (which are single char).
+
+    The patterns for the single-chars and the subjoined letters:
+    """
+    generated_single_chars = "".join(
+        nukta_udaat_yakash_order_list
+        + vowel_order_list
+        + remaining_diacritic_order_list
+    )
+    d_pattern = "([" + generated_single_chars + "]*)"
+
+    subjoin_constructor = "੍"
+    subjoin_consonants = "ਹਰਵਟਤਨਚ"
+    s_pattern = "(" + subjoin_constructor + "[" + subjoin_consonants + "])?"
+
+    """
+    The following regex will capture all sequential diacritics containing at most one subjoined letter.
+        >>> print(regex_match_pattern)
+        '([਼ੵਿੇੈੋੌੁੂਾੀਁੱਂੰਃ]*)(੍[ਹਰਵਟਤਨਚ])?([਼ੵਿੇੈੋੌੁੂਾੀਁੱਂੰਃ]*)'
+    """
+    regex_match_pattern = d_pattern + s_pattern + d_pattern
+
+    """
+    This generates a string of the order in which all diacritics should appear.
+        >>> print(generated_diacritic_order)
+        '਼ਹਰਵਟਤਨਚਿੇੈੋੌੁੂਾੀਁੱਂੰਃ'
+   """
+    generated_diacritic_order = "".join(
+        nukta_udaat_yakash_order_list
+        + [subjoin_constructor]
+        + [subjoin_consonants]
+        + vowel_order_list
+        + remaining_diacritic_order_list
+    )
+
+    def regex_sort_func(match: Match) -> str:
+        """
+        This re-arranges characters in "match" according to a custom sort order "generated_diacritic_order"
+        """
+        if len(match := match.group()) > 1:
```

Comment:
you generally want to avoid using the same variable name in a function as a parameter. renaming the local match to _match is usually a quick and easy fix.

Reply:
👍 Okay, I will do that. What's the underscore indicate?